File: frontend-admin-pro/src/api/subcontract.ts
import { http } from '@/utils/http'

export type SubcontractItemOut = {
  id: number
  sku_id: number
  sku_code: string | null
  sku_name: string | null
  qty: number
  unit_price: number | null
  sent_qty: number
  received_qty: number
  remark: string | null
}

export type SubcontractOut = {
  id: number
  tenant_id: number
  supplier_id: number
  supplier_name: string | null
  code: string
  status: string
  remark: string | null
  created_by: number | null
  created_at: string
  updated_at: string
  items: SubcontractItemOut[]
}

export function listSubcontracts(supplierId?: number, status?: string, offset = 0, limit = 50) {
  return http.get<SubcontractOut[]>('/admin/subcontract', { params: { supplier_id: supplierId, status, offset, limit } })
}

export function getSubcontract(id: number) {
  return http.get<SubcontractOut>(`/admin/subcontract/${id}`)
}

export function createSubcontract(supplierId: number, remark?: string) {
  return http.post<SubcontractOut>('/admin/subcontract', null, { params: { supplier_id: supplierId, remark } })
}

export function addSubcontractItems(id: number, items: { sku_id: number; qty: number; unit_price?: number; remark?: string }[]) {
  return http.post<SubcontractOut>(`/admin/subcontract/${id}/items`, items)
}

export function sendSubcontract(id: number) {
  return http.post<SubcontractOut>(`/admin/subcontract/${id}/send`)
}

export function receiveSubcontract(id: number, itemId: number, qty: number) {
  return http.post<SubcontractOut>(`/admin/subcontract/${id}/receive`, null, { params: { item_id: itemId, qty } })
}

export function settleSubcontract(id: number) {
  return http.post<SubcontractOut>(`/admin/subcontract/${id}/settle`)
}

File: backend/app/schemas/subcontract.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class SubcontractItemIn(BaseModel):
    sku_id: int = Field(ge=1)
    process_id: int | None = Field(default=None, ge=1)
    qty: int = Field(ge=1)
    unit_price: Decimal | None = None
    remark: str | None = None


class SubcontractCreateIn(BaseModel):
    supplier_id: int = Field(ge=1)
    remark: str | None = None


class SubcontractItemOut(BaseModel):
    id: int
    sku_id: int
    sku_code: str | None = None
    sku_name: str | None = None
    process_id: int | None = None
    process_name: str | None = None
    qty: int
    unit_price: Decimal | None = None
    sent_qty: int
    received_qty: int
    remark: str | None = None


class SubcontractOut(BaseModel):
    id: int
    tenant_id: int
    supplier_id: int
    supplier_name: str | None = None
    code: str
    status: str
    remark: str | None = None
    created_by: int | None = None
    created_at: datetime
    updated_at: datetime
    items: list[SubcontractItemOut] = []


class SubcontractSendIn(BaseModel):
    """发料请求：批量发料"""
    sends: list[dict] = Field(description="[{item_id, qty}, ...]")


class SubcontractSendLogOut(BaseModel):
    id: int
    item_id: int
    qty: int
    remark: str | None = None
    sent_by: int | None = None
    sent_at: datetime


class SubcontractReceiveIn(BaseModel):
    """收货请求"""
    item_id: int = Field(ge=1)
    qty: int = Field(ge=1)
    remark: str | None = None


class SubcontractReceiveLogOut(BaseModel):
    id: int
    item_id: int
    qty: int
    remark: str | None = None
    received_by: int | None = None
    received_at: datetime
